### PR TITLE
Fix crash in comparison functions due to incorrect query analysis

### LINKED ISSUE #52057: ClickHouse Server 23.7.1.857 terminated by Null Pointer Access after running a SQL statement
**Describe the bug**
The ClickHouse server terminated by Null Pointer Access after running a SQL statement.

**How to reproduce**
The SQL statement to reproduce:
```sql
SELECT
    x
FROM
(
    SELECT
        (
            WITH x AS
                (
                    SELECT 1 AS x
                )
            SELECT 1 NOT IN (
                    SELECT 1
                ) AS x
            FROM x
            INNER JOIN x USING (x)
        ) AS x
    FROM
    (
        SELECT -1 AS x
    )
) AS x
WHERE (* NOT IN (1)) = 1
```

It can be reproduced on the official docker image. (`clickhouse/clickhouse-server:head` and `clickhouse/clickhouse-server:latest`).

The log:
```
[a75f227bd0b9] 2023.07.12 09:38:23.474449 [ 404 ] <Fatal> BaseDaemon: ########################################
[a75f227bd0b9] 2023.07.12 09:38:23.474510 [ 404 ] <Fatal> BaseDaemon: (version 23.7.1.857 (official build), build id: EBF7F00D257A03E61ABF10B376C4A7BDD0EED2BA, git hash: 9aef39a7888e10995ea85faad559a110c0e22a82) (from thread 48) (query_id: 934c546b-51d2-4538-b58d-49007aeea07c) (query: SELECT
    x
FROM
(
    SELECT
        (
            WITH x AS
                (
                    SELECT 1 AS x
                )
            SELECT 1 NOT IN (
                    SELECT 1
                ) AS x
            FROM x
            INNER JOIN x USING (x)
        ) AS x
    FROM
    (
        SELECT -1 AS x
    )
) AS x
WHERE (* NOT IN (1)) = 1) Received signal Segmentation fault (11)
[a75f227bd0b9] 2023.07.12 09:38:23.479126 [ 404 ] <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Unknown si_code.
[a75f227bd0b9] 2023.07.12 09:38:23.479164 [ 404 ] <Fatal> BaseDaemon: Stack trace: 0x0000000013d197b5 0x000000000a082d64 0x000000000a06a4f7 0x0000000008857fea 0x000000000885758e 0x00000000127633ef 0x0000000012763e62 0x0000000012765159 0x0000000012eec78c 0x0000000014c7c844 0x0000000014ddcf12 0x000000001381ae40 0x000000001380bdf3 0x0000000013809734 0x00000000138a7bf6 0x00000000138a8b04 0x0000000013bd8153 0x0000000013bd422e 0x00000000149f8284 0x0000000014a0f759 0x000000001798e714 0x000000001798f931 0x0000000017b11727 0x0000000017b0f15c 0x00007f8cb5a5a609 0x00007f8cb597f133
[a75f227bd0b9] 2023.07.12 09:38:23.479268 [ 404 ] <Fatal> BaseDaemon: 2. DB::ColumnNullable::compareAt(unsigned long, unsigned long, DB::IColumn const&, int) const @ 0x0000000013d197b5 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479336 [ 404 ] <Fatal> BaseDaemon: 3. ? @ 0x000000000a082d64 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479382 [ 404 ] <Fatal> BaseDaemon: 4. ? @ 0x000000000a06a4f7 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479441 [ 404 ] <Fatal> BaseDaemon: 5. ? @ 0x0000000008857fea in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479506 [ 404 ] <Fatal> BaseDaemon: 6. ? @ 0x000000000885758e in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479570 [ 404 ] <Fatal> BaseDaemon: 7. DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000127633ef in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479635 [ 404 ] <Fatal> BaseDaemon: 8. DB::IExecutableFunction::executeWithoutSparseColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x0000000012763e62 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479710 [ 404 ] <Fatal> BaseDaemon: 9. DB::IExecutableFunction::execute(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x0000000012765159 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479827 [ 404 ] <Fatal> BaseDaemon: 10. DB::ActionsDAG::updateHeader(DB::Block) const @ 0x0000000012eec78c in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479898 [ 404 ] <Fatal> BaseDaemon: 11. DB::FilterTransform::transformHeader(DB::Block, DB::ActionsDAG const*, String const&, bool) @ 0x0000000014c7c844 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.479977 [ 404 ] <Fatal> BaseDaemon: 12. DB::FilterStep::FilterStep(DB::DataStream const&, std::shared_ptr<DB::ActionsDAG> const&, String, bool) @ 0x0000000014ddcf12 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480025 [ 404 ] <Fatal> BaseDaemon: 13. DB::InterpreterSelectQuery::executeWhere(DB::QueryPlan&, std::shared_ptr<DB::ActionsDAG> const&, bool) @ 0x000000001381ae40 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480054 [ 404 ] <Fatal> BaseDaemon: 14. DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::optional<DB::Pipe>) @ 0x000000001380bdf3 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480092 [ 404 ] <Fatal> BaseDaemon: 15. DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&) @ 0x0000000013809734 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480135 [ 404 ] <Fatal> BaseDaemon: 16. DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&) @ 0x00000000138a7bf6 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480207 [ 404 ] <Fatal> BaseDaemon: 17. DB::InterpreterSelectWithUnionQuery::execute() @ 0x00000000138a8b04 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480263 [ 404 ] <Fatal> BaseDaemon: 18. ? @ 0x0000000013bd8153 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480317 [ 404 ] <Fatal> BaseDaemon: 19. DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x0000000013bd422e in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480372 [ 404 ] <Fatal> BaseDaemon: 20. DB::TCPHandler::runImpl() @ 0x00000000149f8284 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480422 [ 404 ] <Fatal> BaseDaemon: 21. DB::TCPHandler::run() @ 0x0000000014a0f759 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480478 [ 404 ] <Fatal> BaseDaemon: 22. Poco::Net::TCPServerConnection::start() @ 0x000000001798e714 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480528 [ 404 ] <Fatal> BaseDaemon: 23. Poco::Net::TCPServerDispatcher::run() @ 0x000000001798f931 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480571 [ 404 ] <Fatal> BaseDaemon: 24. Poco::PooledThread::run() @ 0x0000000017b11727 in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480634 [ 404 ] <Fatal> BaseDaemon: 25. Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000017b0f15c in /usr/bin/clickhouse
[a75f227bd0b9] 2023.07.12 09:38:23.480676 [ 404 ] <Fatal> BaseDaemon: 26. ? @ 0x00007f8cb5a5a609 in ?
[a75f227bd0b9] 2023.07.12 09:38:23.480726 [ 404 ] <Fatal> BaseDaemon: 27. clone @ 0x00007f8cb597f133 in ?
[a75f227bd0b9] 2023.07.12 09:38:23.665405 [ 404 ] <Fatal> BaseDaemon: Integrity check of the executable successfully passed (checksum: C8DB2CC6CEEB111C8AB95A0F4BA5538E)
[a75f227bd0b9] 2023.07.12 09:38:23.665926 [ 404 ] <Fatal> BaseDaemon: Report this error to https://github.com/ClickHouse/ClickHouse/issues
[a75f227bd0b9] 2023.07.12 09:38:23.666124 [ 404 ] <Fatal> BaseDaemon: No settings were changed
```

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

This does not close #52057 but prevents this particular query to affect release builds.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit 34ec39234f875d8309055b196fd083fc0fe74586 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52172/34ec39234f875d8309055b196fd083fc0fe74586/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Functions/FunctionsComparison.h [hunk 1]
```diff
@@ -1112,6 +1112,11 @@ class FunctionComparison : public IFunction
         bool c0_const = isColumnConst(*c0);
         bool c1_const = isColumnConst(*c1);
 
+        /// This is a paranoid check to protect from a broken query analysis.
+        if (c0->isNullable() != c1->isNullable())
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Functions/FunctionsComparison.h:1116]: The whole point was in the fact that the data types are identical, but the columns are not.

[the remark this answers is not in the document]
Comment (novikd) [reply src/Functions/FunctionsComparison.h:1116]: I expect `IColumn::getDataType` to return different `TypeIndex` for different columns.

[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Functions/FunctionsComparison.h:1116]: TypeIndex is related to DataType, not for columns.

**Pull request merged**